enhancements/ingress/custom-route-configuration.md

```diff
@@ -0,0 +1,242 @@
+3. provides a way to have an operator with scope limited permissions read the serving cert/key pairs
+4. allow a cluster-admin to specify a different DNS name.
+
+### Non-Goals
+
+1. provide a way to specify SNI serving certificates for operands.
+2. dictate how an operator must expose itself, the termination policy on a route, or how the secret is used.
+
+## Proposal
+
+```
+type IngressSpec struct {
+    // other fields
+
+    // CustomizeableRoutes is a list of routes that a cluster-admin wants to customize.  It is logically keyed by
+    // .spec.customizeableRoutes[index].{namespace,name}.
+    // To determine the set of possible keys, look at .status.customizeableRoutes where participating operators place
+    // current route status keyed the same way.
+    CustomizeableRoutes []CustomizeableRouteSpec
+}
+
+type IngressStatus struct {
+    // other fields
+
+    // CustomizeableRoutes is where participating operators place the current route status for routes which the cluster-admin
+    // can customize hostnames and serving certificates.
+    // How the operator uses that serving certificate is up to the individual operator.
+    CustomizeableRoutes []CustomizeableRouteStatus
+}
+
+type CustomizeableRouteSpec struct{
+    // namespace is the namespace of the route to customize.  It must be a real namespace.
+    Namespace string
+    // name is the *logical* name of the route to customize.  It does not have to be the actual name of a route resource.
+    // Keep in mind that this is your API for users to customize.  You could later rename the route, but you cannot rename
+    // this name.
+    Name string
+    // desiredHost is the name that a cluster-admin wants to specify
+    DesiredHost string
+    // ServingCertKeyPairSecret is a reference to a secret in namespace/openshift-config that is a kubernetes tls secret.
+    // The serving cert/key pair must match and will be used by the operator to fulfill the intent of serving with this name.
+    // That means it could be embedded into the route or used by the operand directly.
+    // Operands should take care to ensure that if they use passthrough termination, they properly use SNI to allow service
+    // DNS access to continue to function correctly.
+    ServingCertKeyPairSecret SecretNameReference
+
+    // possible future, we could add a set of SNI mappings.  I suspect most operators would not properly handle it today.
+}
+
+type CustomizeableRouteStatus struct{
+    // namespace is the namespace of the route to customize.  It must be a real namespace.
+    Namespace string
+    // name is the *logical* name of the route to customize.  It does not have to be the actual name of a route resource.
+    // Keep in mind that this is your API for users to customize.  You could later rename the route, but you cannot rename
+    // this name.
+    Name string
+    // defaultHost is the normal name of this route.
+    DefaultHost string
```

Comment:
Is this needed?  We don't usually have status fields to say what value we _would_ use if spec didn't specify a value.

Reply:
> Is this needed? We don't usually have status fields to say what value we _would_ use if spec didn't specify a value

Recognizeability.  One mapping is namespace/name.  Another likely mapping is "this is the name of the thing I want to customize".  Displaying both in status allows a user to find names based on either.

Reply:
> Is this needed?

When console is accessed via `console.foo.com`, how does it know which auth and which api name it should use? I understand it would pick `DefaultHost` here and forward the user to the right auth URL or to use the right api URL in the kubeconfig the user can download.

@deads2k is this your understanding?

Reply:
This also means that if you have multiple domains, e.g. `console.public-internet.com` and `console.intranet`, the console would have no way to know the matching auth and api URL. @deads2k is that correct?

@Anandnatraj do we have the requirement for such a "polymorphic" mapping of URLs, depending on which console URL the user accesses?

Reply:
> When console is accessed via `console.foo.com`, how does it know which auth and which api name it should use? I understand it would pick `DefaultHost` here and forward the user to the right auth URL or to use the right api URL in the kubeconfig the user can download.
> 
> @deads2k is this your understanding?

I would probably choose the first `.status.currentHosts` value.



> This also means that if you have multiple domains, e.g. `console.public-internet.com` and `console.intranet`, the console would have no way to know the matching auth and api URL. @deads2k is that correct?

I don't think this makes that worse one way or the other.

Reply:
> I don't think this makes that worse one way or the other.

the choice is important if not all URLs can be accessed from the internet.

Reply:
> > I don't think this makes that worse one way or the other.
> 
> the choice is important if not all URLs can be accessed from the internet.

The route API doesn't know which it is either.

Reply:
Not talking about the route, but about console and oauth-server choosing the right URL.